server/CGameHandler.cpp

```diff
@@ -6222,6 +6391,12 @@ CasualtiesAfterBattle::CasualtiesAfterBattle(const CArmedInstance * _army, Battl
 				StackLocation sl(army, st->slot);
 				newStackCounts.push_back(TStackAndItsNewCount(sl, st->count));
 			}
+			else if (st->count > army->getStackCount(st->slot) && !st->hasBonusOfType(Bonus::SOUL_STEAL, 1))
```

Comment:
Sorry, missed that comment.
Then you should use "resurrected" field of CStack like low level resurrection do (more over you designed SOUL_STEAL as "overheal"). No checks here.